x/ecocredit/server/marketplace/features/msg_buy_direct.feature

```diff
@@ -0,0 +1,324 @@
+
+    Scenario: buyer bank balance is less than total cost (multiple buy orders)
+      Given alice created two sell orders each with quantity "10" and ask amount "10"
+      And bob has a bank balance with amount "150"
+      When bob attempts to buy credits in two orders each with quantity "10" and bid amount "10"
+      Then expect the error "orders[1]: quantity: 10, ask price: 10regen, total price: 100regen, bank balance: 50regen: insufficient funds"
+
+  Rule: The buyer must provide a bid price greater than or equal to the ask price
+
+    Background:
+      Given a credit type
+      And alice created a sell order with ask amount "10"
+      And bob has a bank balance with amount "100"
+
+    Scenario Outline: bid price greater than or equal to ask price
+      When bob attempts to buy credits with quantity "10" and bid amount "<bid-amount>"
+      Then expect no error
+
+      Examples:
+        | description  | bid-amount |
+        | greater than | 20         |
+        | equal to     | 10         |
+
+    Scenario: bid price less than ask price
+      When bob attempts to buy credits with quantity "10" and bid amount "5"
+      Then expect the error "orders[0]: ask price: 10regen, bid price: 5regen, insufficient bid price: invalid request"
+
+  Rule: The buyer must provide a quantity less than or equal to the sell order quantity
+
+    Background:
+      Given a credit type
+      And alice created a sell order with quantity "10"
+      And bob has a bank balance with amount "150"
+
+    Scenario Outline: quantity less than or equal to sell order quantity
+      When bob attempts to buy credits with quantity "<quantity>"
+      Then expect no error
+
+      Examples:
+        | description | quantity |
+        | less than   | 5        |
+        | equal to    | 10       |
+
+    Scenario: quantity more than sell order quantity
+      When bob attempts to buy credits with quantity "15"
+      Then expect the error "orders[0]: requested quantity: 15, sell order quantity 10: invalid request"
+
+  Rule: The number of decimal places in quantity must be less than or equal to the credit type precision
+
+    Background:
+      Given a credit type with precision "6"
+
+    Scenario Outline: quantity decimal places less than or equal to precision
+      Given alice created a sell order with quantity "<quantity>"
+      When bob attempts to buy credits with quantity "<quantity>"
+      Then expect no error
+
+      Examples:
+        | description | quantity |
+        | less than   | 9.12345  |
+        | equal to    | 9.123456 |
+
+    Scenario: quantity decimal places more than precision
+      Given alice created a sell order with quantity "9.1234567"
+      When bob attempts to buy credits with quantity "9.1234567"
+      Then expect the error "orders[0]: quantity: 9.1234567, credit type precision: 6: invalid request"
```

Comment:
i feel like users won't immediately understand why this is an invalid request, i think we could specify a bit more

Reply:
https://github.com/regen-network/regen-ledger/pull/1122/commits/023fd694b90e82d70b1646d70f3e20dd5043d5fe